presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.jdbc;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.airlift.units.Duration;
+import io.prestosql.spi.connector.ColumnMetadata;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.ConnectorTableMetadata;
+import io.prestosql.spi.connector.SchemaTableName;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class CachingJdbcClient
+        extends ForwardingJdbcClient
+{
+    private final Cache<ColumnsCacheKey, List<JdbcColumnHandle>> columnsCache;
+    private final Cache<TableHandleCacheKey, Optional<JdbcTableHandle>> tableHandleCache;
+    private final Cache<JdbcIdentity, Set<String>> schemaNamesCache;
+    private final Cache<TableNamesCacheKey, List<SchemaTableName>> tableNamesCache;
+
+    private final JdbcClient delegate;
+    private final boolean cacheMissing;
+
+    @Inject
+    public CachingJdbcClient(@StatsCollecting JdbcClient delegate, BaseJdbcConfig config)
+    {
+        this(delegate, config.getMetadataCachingTtl(), config.isCacheMissing());
+    }
+
+    public CachingJdbcClient(JdbcClient delegate, Duration metadataCachingTtl, boolean cacheMissing)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(metadataCachingTtl.toMillis(), TimeUnit.MILLISECONDS);
+        columnsCache = cacheBuilder.build();
+        tableHandleCache = cacheBuilder.build();
+        schemaNamesCache = cacheBuilder.build();
+        tableNamesCache = cacheBuilder.build();
+        this.cacheMissing = cacheMissing;
+    }
+
+    @Override
+    protected JdbcClient getDelegate()
+    {
+        return delegate;
+    }
+
+    @Override
+    public Set<String> getSchemaNames(JdbcIdentity identity)
+    {
+        Set<String> schemas = schemaNamesCache.getIfPresent(identity);
+        if (schemas != null) {
+            return schemas;
+        }
+
+        schemas = getDelegate().getSchemaNames(identity);
+        schemaNamesCache.put(identity, schemas);
```

Comment:
Use `CacheLoader`. This enforces that cached value is a function of the cache key, so that we won't be able to make a mistake that eg cached info leaks between users.